src/dbnode/ts/writes/write_batch.go

```diff
@@ -102,14 +128,29 @@ func (b *writeBatch) Reset(
 	batchSize int,
 	ns ident.ID,
 ) {
-	var writes []BatchWrite
+	// Preallocate slightly more when not using initialBatchSize.
+	adaptiveBatchCap := int(float32(batchSize) * preallocateBatchCoeff)
+
 	if batchSize > cap(b.writes) {
-		writes = make([]BatchWrite, 0, batchSize)
+		batchCap := batchSize
+		if b.adaptiveSize {
+			batchCap = adaptiveBatchCap
+		}
+		b.writes = make([]BatchWrite, 0, batchCap)
```

Comment:
nit: I think it would be very useful to have a gauge for current allocated batch capacity. It could be valuable in cases when, e.g. we receive one huge batch and subsequent batches are much smaller. The memory usage will probably be increased after this even if we are receiving smaller batches so this gauge could help us to identify such cases.

Reply:
How would this gauge work? I guess we would increment it on line 139, but when would we decrement it?

Reply:
It could be also a counter for now, I was thinking that maybe in future we will update this to also shrink batch capacity so gauge makes sense from this point of view.